controllers/cluster_controller.go

```diff
@@ -231,30 +239,46 @@ func (r *ClusterReconciler) reconcileDelete(ctx context.Context, cluster *cluste
 	return ctrl.Result{}, nil
 }
 
-// listChildren returns a list of MachineDeployments, MachineSets, and Machines than have an owner reference to cluster
-func (r *ClusterReconciler) listChildren(ctx context.Context, cluster *clusterv1.Cluster) ([]runtime.Object, error) {
-	logger := r.Log.WithValues("cluster", cluster.Name, "namespace", cluster.Namespace)
+type clusterDescendants struct {
+	machineDeployments clusterv1.MachineDeploymentList
+	machineSets        clusterv1.MachineSetList
+	machines           clusterv1.MachineList
+}
+
+// empty returns true if all the lists are empty.
+func (c *clusterDescendants) empty() bool {
+	return len(c.machineDeployments.Items) == 0 && len(c.machineSets.Items) == 0 && len(c.machines.Items) == 0
+}
+
+// listAllClusterDescendants returns a list of all MachineDeployments, MachineSets, and Machines for the cluster.
+func (r *ClusterReconciler) listAllClusterDescendants(ctx context.Context, cluster *clusterv1.Cluster) (clusterDescendants, error) {
```

Comment:
```suggestion
func (r *ClusterReconciler) listDescendants(ctx context.Context, cluster *clusterv1.Cluster) (clusterDescendants, error) {
```

Given that this is the cluster reconciler, `cluster` can probably be omitted from the name